### ISSUE TITLE
Is it possible to use/support OIDC authorization with web hooks?

### ISSUE THREAD

**Issue by MohammedNoureldin** [external-secrets/external-secrets#2813]
**Is your feature request related to a problem? Please describe.**

I am trying to mount some secrets from REST-API that requires authentication by the mediation of an identity provider. So in other words I am using the web hook provider and have to be able to somehow send a JWT in the authorization header. For that, External Secrets has to be able to send a request to the identity provider with the client ID and client secret to fetch the JWT, which will then be sent in the header to the API.

The step of contacting the IdP to fetch the JWT is missing.

**Describe the solution you'd like**

We should be somehow be able to, for example, specify the token endpoint in the configuration of the `SecretStore`, to which External Secrets will be contacting to fetch a JWT when needed (if expired). And then ESO will put the returned JWT in the headers.

I am not sure if you have other alternatives for this or any better approach.

**Comment by moolen** [comment]
Hey @MohammedNoureldin :wave:, this is a reasonable request! 
I think it would belong into the SecretStore spec where a user would specify a token endpoint, id token `secretRef` (or `serviceAccountRef` in case ), eso would then exchange the token and pass it along the other secrets.

I guess the hardest part would be designing an API that supports all the different flows. 


As an alternative, depending on context, it may be an option to have a CronJob which exchanges the token and stores it as `Kind=Secret`. ESO would then be able to pick it up and use it in a `Authentication` header. Depending on token `exp` this may be feasible or not.

**Comment by MohammedNoureldin** [comment]
Hi @moolen!

> I guess the hardest part would be designing an API that supports all the different flows.

I would say, not all flows should necessary be implemented together. In most cases, because ESO is a backend service, the needed flow will be ***Client Credentials Grant***, where the admin passes the client ID and secret, and ESO will fetch accordingly an access token and put it in the request head that will be sent to the API. Maybe this one can be implemented first because, as said, it will cover most use-cases?

Exactly, SecretStore and ClusterSecretStore spec is a good place for such feature. Maybe ESO will not pass the token automatically, but expose the fetched token as a variable. Just like secretRef, or maybe in addition to it.

To not affect the secretRef, and to make this more useful and generic, maybe a `preRequest` field can be added, which is nothing more than a URL, header and a body. So the user can specify whatever he needs in this preRequest, and then the fetched result can be accessed using a `{variable_name}` he specifies, just like secretRef. If this is complexer, I think a simpler version to just fetch the token and pass it automatically is also very OK. What would you say?

Thanks for Cron job suggestion, it is a clever one. Still, having a ESO native solution would be nice.

**Comment by moolen** [comment]
> not all flows should necessary be implemented together

Agree, what i wanted to say is that it should be designed in a way that we're able to add more flows/features later on. Like e.g. we should eventually be able to read `Kind=Secret` resources or offer a way to use the `TokenRequest` API so a user can forward these to the exchange endpoint.

I really like the `preRequest` idea, that sounds good to me! Maybe it should be an array, so the user is able to execute multiple `preRequest` tasks before calling the webhook target.

**Comment by dholeshu** [comment]
Hi !
Has there been any progress on this idea, we would really appreciate that feature as well 👍🏻

**Comment by gusfcarvalho** [comment]
Hi @dholeshu . Just to be clear - the fact that we maintainers respond to feature requests validating them as useful/not duplicate/etc, does not point to any SLAs on implementing them 😄.

If you need this feature, you should try to implement it yourself. We'd be more than happy to review PRs

**Comment by gberche-orange** [comment]
See workaround  with a cronjob which generates a secret containing a bearer token, and then reference it in the Authorization header at https://github.com/external-secrets/external-secrets/issues/3871#issuecomment-3511119930

**Comment by evrardj-roche** [comment]
This issue has been marked as Stale and will be closed soon if there is no activity. Please update the issue or engage with it to keep it open. It is fine to re-open this issue in the future if you want to work on this.

**Comment by evrardj-roche** [comment]
Hello. Thank you for taking the time to open this issue and contribute to the project.
We truly appreciate community involvement and the effort that goes into reporting issues and suggesting improvements.

We're closing this (Stale) issue as part of a maintenance effort to keep our issue tracker focused and actionable.
Issues that have been inactive for an extended period can create clutter, make it harder to prioritize current work, and distract maintainers from issues that have active contributors ready to work on them.

This doesn't mean your issue isn't valid or important.
If you or anyone else is still affected by this issue and willing to contribute toward a solution (whether through discussion, testing, or code),
please feel free to comment here, reopen the issue, and mention you want to be assigned on this issue.

Alternatively, you can directly submit a PR with `Closes <PR NUMBER>` in its content.

We're happy to reopen or revisit this if there's renewed interest and someone willing to champion it forward.

Thanks again for being part of our community!
